File: Tests.VisualBasic/Defender/GuardTests.vb
Imports Defender
Imports Defender.Exceptions
Imports Xunit

Public Class GuardTests
	<Theory>
	<InlineData(1, 1, False)>
	<InlineData(1, 2, True)>
	<InlineData(2, 1, True)>
	<InlineData("a"c, "a"c, False)>
	<InlineData("a"c, "b"c, True)>
	<InlineData("b"c, "a"c, True)>
	<InlineData("hi", "hi", False)>
	<InlineData("hi", "no", True)>
	<InlineData("no", "hi", True)>
	Public Sub Equal(Of T)(Value As T, Other As T, Throws As Boolean)
		If Throws Then
			Assert.Throws(Of ArgumentNotEqualException)(Sub() Guard.Equal(Value, "Value", Other))
		Else
			Guard.Equal(Value, "Value", Other)
		End If
	End Sub

	<Theory>
	<InlineData(0, 1, False)>
	<InlineData(1, 1, False)>
	<InlineData(2, 1, True)>
	Public Sub GreaterThan(Value As Integer, Lower As Integer, Throws As Boolean)
		If Throws Then
			Assert.ThrowsAny(Of ArgumentException)(Sub() Guard.GreaterThan(Value, "Value", Lower))
		Else
			Guard.GreaterThan(Value, "Value", Lower)
		End If
	End Sub

	<Theory>
	<InlineData(0, 1, True)>
	<InlineData(1, 1, False)>
	<InlineData(2, 1, False)>
	Public Sub LesserThan(Value As Integer, Upper As Integer, Throws As Boolean)
		If Throws Then
			Assert.ThrowsAny(Of ArgumentException)(Sub() Guard.LesserThan(Value, "Value", Upper))
		Else
			Guard.LesserThan(Value, "Value", Upper)
		End If
	End Sub

	<Theory>
	<InlineData(Nothing, True)>
	<InlineData(1, False)>
	<InlineData(1.5, True)>
	<InlineData("a"c, True)>
	<InlineData("hi", True)>
	Public Sub OfType(Value As Object, Throws As Boolean)
		If Throws Then
			Assert.Throws(Of ArgumentNotTypeException)(Sub() Guard.OfType(Of Integer)(Value, "Value"))
		Else
			Guard.OfType(Of Integer)(Value, "Value")
		End If
	End Sub

	<Theory>
	<InlineData(New Integer() {}, True)>
	<InlineData(New Integer() {1, 2, 3}, False)>
	Public Sub NotEmpty(Value As Integer(), Throws As Boolean)
		If Throws Then
			Assert.Throws(Of ArgumentEmptyException)(Sub() Guard.NotEmpty(Value, "Value"))
		Else
			Guard.NotEmpty(Value, "Value")
		End If
	End Sub

	<Theory>
	<InlineData("", False)>
	<InlineData(Nothing, True)>
	Public Sub NotNull(Value As String, Throws As Boolean)
		If Throws Then
			Assert.Throws(Of ArgumentNullException)(Sub() Guard.NotNull(Value, "Value"))
		Else
			Guard.NotNull(Value, "Value")
		End If
	End Sub

	<Theory>
	<InlineData(-1, True)>
	<InlineData(0, False)>
	<InlineData(1, False)>
	<InlineData(15, False)>
	<InlineData(16, True)>
	Public Sub Valid(Value As Integer, Throws As Boolean)
		If Throws Then
			Assert.Throws(Of ArgumentNotValidException)(Sub() Guard.Valid(CType(Value, ConsoleColor), "Value"))
		Else
			Guard.Valid(CType(Value, ConsoleColor), "Value")
		End If
	End Sub

	<Theory>
	<InlineData(0, 1, 5, True)>
	<InlineData(1, 1, 5, False)>
	<InlineData(2, 1, 5, False)>
	<InlineData(4, 1, 5, False)>
	<InlineData(5, 1, 5, False)>
	<InlineData(6, 1, 5, True)>
	Public Sub Within(Value As Integer, Lower As Integer, Upper As Integer, Throws As Boolean)
		If Throws Then
			Assert.Throws(Of ArgumentNotWithinException)(Sub() Guard.Within(Value, "Value", Lower, Upper))
		Else
			Guard.Within(Value, "Value", Lower, Upper)
		End If
	End Sub
End Class

File: Tests/Defender/GuardTests.cs
using System;
using Xunit;
using Langly;

namespace Defender {
	public class GuardTests {
		[Theory]
		[InlineData('a', 'a', true)]
		[InlineData('a', 'b', false)]
		[InlineData('b', 'a', false)]
		public void Equal_Char(Char value, Char other, Boolean succeeds) {
			if (succeeds) {
				Guard.Equal(value, nameof(value), other);
			} else {
				Assert.Throws<ArgumentNotEqualException>(() => Guard.Equal(value, nameof(value), other));
			}
		}

		[Theory]
		[InlineData(1, 1, true)]
		[InlineData(1, 2, false)]
		[InlineData(2, 1, false)]
		public void Equal_Int32(Int32 value, Int32 other, Boolean succeeds) {
			if (succeeds) {
				Guard.Equal(value, nameof(value), other);
			} else {
				Assert.Throws<ArgumentNotEqualException>(() => Guard.Equal(value, nameof(value), other));
			}
		}
		[Theory]
		[InlineData("hi", "hi", true)]
		[InlineData("hi", "no", false)]
		[InlineData("no", "hi", false)]
		public void Equal_String(String value, String other, Boolean succeeds) {
			if (succeeds) {
				Guard.Equal(value, nameof(value), other);
			} else {
				Assert.Throws<ArgumentNotEqualException>(() => Guard.Equal(value, nameof(value), other));
			}
		}

		[Theory]
		[InlineData(0, 1, false)]
		[InlineData(1, 1, false)]
		[InlineData(2, 1, true)]
		public void GreaterThan_Int32(Int32 value, Int32 lower, Boolean succeeds) {
			if (succeeds) {
				Guard.GreaterThan(value, nameof(value), lower);
			} else if (value == lower) {
				Assert.Throws<ArgumentEqualException>(() => Guard.GreaterThan(value, nameof(value), lower));
			} else {
				Assert.Throws<ArgumentLesserThanException>(() => Guard.GreaterThan(value, nameof(value), lower));
			}
		}

		[Theory]
		[InlineData(0, 1, false)]
		[InlineData(1, 1, true)]
		[InlineData(2, 1, true)]
		public void GreaterThanOrEqual_Int32(Int32 value, Int32 lower, Boolean succeeds) {
			if (succeeds) {
				Guard.GreaterThanOrEqual(value, nameof(value), lower);
			} else {
				Assert.Throws<ArgumentLesserThanException>(() => Guard.GreaterThanOrEqual(value, nameof(value), lower));
			}
		}

		[Theory]
		[InlineData(0, 1, true)]
		[InlineData(1, 1, false)]
		[InlineData(2, 1, false)]
		public void LesserThan_Int32(Int32 value, Int32 upper, Boolean succeeds) {
			if (succeeds) {
				Guard.LesserThan(value, nameof(value), upper);
			} else if (value == upper) {
				Assert.Throws<ArgumentEqualException>(() => Guard.LesserThan(value, nameof(value), upper));
			} else {
				Assert.Throws<ArgumentGreaterThanException>(() => Guard.LesserThan(value, nameof(value), upper));
			}
		}

		[Theory]
		[InlineData(0, 1, true)]
		[InlineData(1, 1, true)]
		[InlineData(2, 1, false)]
		public void LesserThanOrEqual_Int32(Int32 value, Int32 upper, Boolean succeeds) {
			if (succeeds) {
				Guard.LesserThanOrEqual(value, nameof(value), upper);
			} else {
				Assert.Throws<ArgumentGreaterThanException>(() => Guard.LesserThanOrEqual(value, nameof(value), upper));
			}
		}

		[Theory]
		[InlineData(null, false)]
		[InlineData(1, true)]
		[InlineData(1.5, false)]
		[InlineData('a', false)]
		[InlineData("hi", false)]
		public void OfType_Int32(Object value, Boolean succeeds) {
			if (succeeds) {
				Guard.OfType<Int32>(value, nameof(value));
			} else {
				Assert.Throws<ArgumentNotTypeException>(() => Guard.OfType<Int32>(value, nameof(value)));
			}
		}

		[Theory]
		[InlineData(new Int32[] { }, false)]
		[InlineData(new Int32[] { 1, 2, 3 }, true)]
		public void NotEmpty_Array(Int32[] value, Boolean succeeds) {
			if (succeeds) {
				Guard.NotEmpty(value, nameof(value));
			} else {
				Assert.Throws<ArgumentEmptyException>(() => Guard.NotEmpty(value, nameof(value)));
			}
		}

		[Theory]
		[InlineData(null, false)]
		[InlineData("", true)]
		public void NotNull_Object(Object value, Boolean succeeds) {
			if (succeeds) {
				Guard.NotNull(value, nameof(value));
			} else {
				Assert.Throws<Langly.ArgumentNullException>(() => Guard.NotNull(value, nameof(value)));
			}
		}

		[Theory]
		[InlineData(-1, false)]
		[InlineData(0, true)]
		[InlineData(1, true)]
		[InlineData(15, true)]
		[InlineData(16, false)]
		public void Valid_ConsoleColor(Int32 value, Boolean succeeds) {
			if (succeeds) {
				Guard.Valid((ConsoleColor)value, nameof(value));
			} else {
				Assert.Throws<ArgumentNotValidException>(() => Guard.Valid((ConsoleColor)value, nameof(value)));
			}
		}

		[Theory]
		[InlineData(0, 1, 5, false)]
		[InlineData(1, 1, 5, true)]
		[InlineData(2, 1, 5, true)]
		[InlineData(3, 1, 5, true)]
		[InlineData(4, 1, 5, true)]
		[InlineData(5, 1, 5, true)]
		[InlineData(6, 1, 5, false)]
		public void Within_Int32(Int32 value, Int32 lower, Int32 upper, Boolean succeeds) {
			if (succeeds) {
				Guard.Within(value, nameof(value), lower, upper);
			} else {
				Assert.Throws<ArgumentNotWithinException>(() => Guard.Within(value, nameof(value), lower, upper));
			}
		}

		[Theory]
		[InlineData(0, 1, 5, false)]
		[InlineData(1, 1, 5, true)]
		[InlineData(2, 1, 5, true)]
		[InlineData(3, 1, 5, true)]
		[InlineData(4, 1, 5, true)]
		[InlineData(5, 1, 5, true)]
		[InlineData(6, 1, 5, false)]
		public void Within_UInt32(UInt32 value, UInt32 lower, UInt32 upper, Boolean succeeds) {
			if (succeeds) {
				Guard.Within(value, nameof(value), lower, upper);
			} else {
				Assert.Throws<ArgumentNotWithinException>(() => Guard.Within(value, nameof(value), lower, upper));
			}
		}
	}
}
